### PR TITLE
Add CNP

### LINKED ISSUE #25: Add Conditional Neural Processes (CNP)
Paper: https://arxiv.org/abs/1807.01613

### PR DESCRIPTION
**Pull request by AnirudhM1**
closes #25

### Example:

```python
from cnp import CNP, Aggregator

class Encoder(linen.Module):
    @linen.compact
    def __call__(self, x, y, train=True):
        # x : (n_classes * shots, 20)
        # y : (n_classes * shots, 1)
        
        y = y.squeeze() # (n_classes, )
        x = linen.Dense(32)(x)
        x = linen.relu(x)
        x = linen.Dense(64)(x)
        x = linen.relu(x)
        r = linen.Dense(32)(x) # (n_classes * shots, 32)

        aggregates = Aggregator.classification(r, y, ways)
        return aggregates # (ways * 32, )
```

```python
class Decoder(linen.Module):
    @linen.compact
    def __call__(self, r, x, train=True):
        # r : (ways * 32, )
        # x : (batch_size, 20)

        x = jax.vmap(lambda a, b: jnp.concatenate([a, b]), in_axes=(0, None))(x, r) # (batch_size, 20 + ways * 32)
        x = linen.Dense(32)(x)
        x = linen.relu(x)
        x = linen.Dense(64)(x)
        x = linen.relu(x)
        x = linen.Dense(32)(x)
        x = linen.relu(x)
        x = linen.Dense(ways)(x)
        return x
```
```python
key = jax.random.PRNGKey(0)
encoder_init_key, decoder_init_key = jax.random.split(key)

encoder = Encoder()
tx = optax.adam(1e-3)
params = encoder.init(encoder_init_key, jnp.ones((1, 20)), jnp.ones((1, 1)))
encoder_state = TrainState.create(apply_fn=encoder.apply, params=params, tx=tx)

decoder = Decoder()
params = decoder.init(decoder_init_key, jnp.ones((ways*32,)), jnp.ones((1, 20)))
decoder_state = TrainState.create(apply_fn=decoder.apply, params=params, tx=tx)
```

```python
n_epochs = 100
task_count=50
for epoch in range(n_epochs):
    # Sample and train on a batch of tasks
    # for task in range(task_count):
    tasks = sample_batch(meta_train_kloader, task_count)

    encoder_state, decoder_state, loss, metrics = CNP.meta_train_step(encoder_state, decoder_state, tasks, loss_fn=cross_entropy, metrics=(accuracy, ))
        
    print('Epoch  % 2d Loss: %2.5e Avg Acc: %2.5f'%(epoch+1, loss, metrics[0]))
    display.clear_output(wait=True)
```

cc: @veds12

### CODE REVIEW

**Comment by AnirudhM1** [comment]
A standard CNP model contains 3 parts:
- Encoder: **r<sub>i</sub> = h(x<sub>i</sub>, y<sub>i</sub> ;  θ)**
- Aggregator: **r = r<sub>1</sub> ⊕ r<sub>2</sub>⊕ ... ⊕ r<sub>n</sub>**
- Decoder: **logits = g(x, y ; Φ)**

For now I have considered the aggregator part to be included in the encoder. This is because the aggregator can be very different for different tasks. That said, I can include some common aggregators (regression, classification etc) and give an option to pass in a parameter to use one of the common aggregators. If this is being done, the exact function definition has to be discussed.

Any thoughts on this? @veds12

**Comment by veds12** [comment]
@AnirudhM1 a few comments:

- I think at this stage it would be good to have a separate module for different NNs (MLPs, LSTMs). So for encoder and decoder you can just call this module
- We can have a separate module for different types of aggregators, since these will be reused in other algorithms as well
- Please also add proper docstrings to the code

**Comment by veds12** [comment]
@AnirudhM1 any updates on this?

**Comment by AnirudhM1** [comment]
@veds12 Sorry for the delay

> We can have a separate module for different types of aggregators, since these will be reused in other algorithms as well

I have added a utility class which has functions for some common aggregation algorithms.
These functions can be called from inside the `Encoder class`

> Please also add proper docstrings to the code

I have added docstrings to the main code (cnp.py). The above code in the description of this PR is just an example of how to use the interface.


> I think at this stage it would be good to have a separate module for different NNs (MLPs, LSTMs). So for encoder and decoder you can just call this module

Can you please elaborate a bit further on this?
Currently, this model works with an arbitrary architecture for encoder and decoder

**Comment by veds12** [comment]
> Can you please elaborate a bit further on this?
> Currently, this model works with an arbitrary architecture for encoder and decoder

I mean having a separate file from which different architectures (MLP, RNNs, CNNs) can be loaded.
